libfwupd/fwupd-client.c

```diff
@@ -485,19 +419,19 @@ fwupd_client_get_device_by_id (FwupdClient *client,
 }
 
 /**
- * fwupd_client_get_updates:
+ * fwupd_client_get_upgrades_all:
  * @client: A #FwupdClient
  * @cancellable: the #GCancellable, or %NULL
  * @error: the #GError, or %NULL
  *
  * Gets all the devices with known updates.
  *
- * Returns: (element-type FwupdResult) (transfer container): results
+ * Returns: (element-type FwupdDevice) (transfer container): results
```

Comment:
Since there is also a `fwupd_client_get_upgrades`, I think it's still confusing to have `fwupd_client_get_upgrades_all`.  How about instead combining them and allowing an optional argument for which devices to get upgrades for?  If no argument get all the devices, if argument get the one you are interested in.

Reply:
I agree it's a bit confusing. With your proposal we'd have to return a `FwupdDevice` array for each request, and for something like `fwupdmgr get-downgrades` where we show a list it's quite wasteful to construct and encode the same FuDevice for each line of the releases. We'd also then have to return all releases for a FuDevice which is wasteful in the other way when we just care about the newest available release for each device in the gnome-software case.

Counter-proposal: How about we have `GetUpgradesForDevice(device_id)` returning an array of releases, `GetDowngradesForDevice(device_id)` (also releases) and `GetUpgrades()` which returns an array of devices (each with just one release set).

Reply:
or, we just nuke `GetUpgradesAll` and get the client to call `GetDevices()`, then `GetUpgrades()` on each one. Maybe the IPC overhead / round-trip isn't such a big deal anyway.

Reply:
I like the second approach more.  I can't imagine the overhead will actually be that bad.  You can also simplify it by clients not calling `GetUpgrades` on devices not marked upgradable.

Reply:
I agree with you. I've pushed https://github.com/hughsie/fwupd/commit/902df9bbcd509c3786cbf91b9ad55e3209cb6a16 that removes it, I'll merge the commits next time I rebase.